### ISSUE TITLE
Support suppression-xpath element in SuppressionLoader

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4421]
SuppressionLoader should be able to parse ```suppression-xpath``` element and its attributes in accordance with https://github.com/checkstyle/checkstyle/issues/4364. In the scope of this issue it is required to create basic implementation.

**This issue is blocked by #4422.**

**Comment by MEZk** [comment]
Part 1
1) Add the set of TreeWalkerFilters to SuppressionLoader
2) Extract [this code](https://github.com/MEZk/checkstyle/blob/f244f7f1e2fe49d2fcda9cd5c901d751769f73e8/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionsLoader.java#L85-L102) to a method.
3) Add another if block to check ```suppress-xpath```.
4) Add a method which will return XpathFilter based on the information taken from suppress-xpath attributes 
5) Add XpathFilter to the set.
6) Add a new method to load a set of TreeWalkerFilters from xml.

Part 2
1) Add a new class SuppressionXpathFilter (similar to SuppressionFilter). The filter should implement TreeWalkerFilter, and ExternalResourceHolder interface. The filter should also extend AutomaticBean.
2) Override finishLocalSetup to load the set of XpathFilter objects via SuppressionLoader.

**Comment by MEZk** [comment]
@timurt 
Any progress?

**Comment by timurt** [comment]
@MEZk 
Yes, my commit [here](https://github.com/timurt/checkstyle/commit/27c4cedb3bb478f0c3117623b405391e98891b23)

**Comment by MEZk** [comment]
@timurt 
Changes look OK. We are waiting for https://github.com/checkstyle/checkstyle/pull/4923 .

Start Working on part 2. When we finish https://github.com/checkstyle/checkstyle/pull/4923 we will merge two parts as separate PR.

**Comment by romani** [comment]
issue is not closed as it is referenced in code as "till ...".

**Comment by rnveach** [comment]
@MEZk @timurt @romani 
> issue is not closed as it is referenced in code as "till ...".

https://github.com/checkstyle/checkstyle/blob/9c43a1d1663746b7cd2357e03bfede9f6cedf9ec/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java#L333

What is required to remove this and complete the issue? I see nothing in this issue/PR stating what we need to do. Are we waiting on another Issue?
PR Notes: https://github.com/checkstyle/checkstyle/pull/4992#discussion_r135298611

**Comment by MEZk** [comment]
@rnveach @romani 
Issue is not finished. We cannot remove the comment. 
1) We need to add the documentation for the new filter. The documentation should explain what checks we support now and the limitations of xpath filter.
2) We need to use SuppressionXpathFilter in our config.
3) We need to replace the suppressions in out suppressions.xml with ```suppress-xpath``` elements where it is possible.

**Comment by rnveach** [comment]
> 3. We need to replace the suppressions in out suppressions.xml with suppress-xpath elements where it is possible.

This is already another issue, https://github.com/checkstyle/checkstyle/issues/4991.

**Comment by MEZk** [comment]
@rnveach 
> This is already another issue, #4991.

Rigth.

@romani 
If I do not receive response from @timurt in a few days, I will continue the work on my own. Is it OK?

**Comment by romani** [comment]
Sure, thanks a lot.

**Comment by timurt** [comment]
@romani @rnveach @MEZk 
Sorry for long delay, I created PR, could you please review

**Comment by rnveach** [comment]
@MEZk @timurt Confirm everything is done for this issue.

> > 3. We need to replace the suppressions in out suppressions.xml with suppress-xpath elements where it is possible.

> This is already another issue, #4991.

There were some changes for this in the last PR. Is Issue #4991 now not needed?

**Comment by MEZk** [comment]
@timurt 
https://github.com/checkstyle/checkstyle/pull/5390/files is a great update. Thanks.

@romani @timurt 
Why did you decided to have separate file (suppressions-xpath.xml)? The basic idea is to allow the users to have the suppression declarations for both old and new model in one file. The dtd schema allows to do this. Did you do this to show the migration process to avoid the mess in one suppression file?

@rnveach 
Yes, https://github.com/checkstyle/checkstyle/issues/4991 was closed. We convert other suppressions when we deal with
1) https://github.com/checkstyle/checkstyle/issues/4530 (will ease the migration process)
2) https://github.com/checkstyle/checkstyle/issues/4830 (will cover all other Checks).

**Comment by MEZk** [comment]
@rnveach @romani @timurt 
I'm closing this issue. The next steps should be addressed in https://github.com/checkstyle/checkstyle/issues/4530 and https://github.com/checkstyle/checkstyle/issues/4830.

**Comment by rnveach** [comment]
> Why did you decided to have separate file (suppressions-xpath.xml)? 
> #4991 was closed. We will convert other suppressions when we deal with

@MEZk please expand all issues so it is clear on the specifics on what is supposed to be done in each issue so there is no more misunderstandings.

**Comment by romani** [comment]
> Why did you decided to have separate file (suppressions-xpath.xml)? 

Looks like lost a context of steps to took to make xpath completed. But it looked so natural, so it did not make rad flag for me during review.

> Did you do this to show the migration process to avoid the mess in one suppression file?

lets keep this file separate till complete xpath implementation. The very last update should simple merge of both suppressions in one file.
